### PR TITLE
Add lean lexer

### PR DESCRIPTION
**Pull request by pcapriotti**
Simple [lean](https://leanprover.github.io) lexer, mostly copied from [the one in pygments](https://bitbucket.org/birkenfeld/pygments-main/src/default/pygments/lexers/theorem.py).

### CODE REVIEW

**Comment by pyrmont** [comment]
@pcapriotti Sorry it's taken a while to get around to this :( Will have a look in the next couple of days and get back to you!

File: lib/rouge/lexers/lean.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+module Rouge
+  module Lexers
+    class Lean < RegexLexer
+      title "lean"
+      desc 'Lean (leanprover.github.io)'
+      tag 'lean'
+      mimetypes 'text/x-lean'
+
+      keywords = Set.new %w(
```
Comment (pyrmont) [inline lib/rouge/lexers/lean.rb]: Can you wrap these local variables in memoised methods? So for example:

```ruby
def self.keywords
  @keywords ||= Set.new %w(...)
end
```

This helps a little with overall performance.

Within the respective rules, it means that calls to `keywords` should be replaced with `self.class.keywords`.

File: lib/rouge/lexers/lean.rb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+      name_start = "A-Za-z_\u03b1-\u03ba\u03bc-\u03fb\u1f00-\u1ffe\u2100-\u214f"
+      name_mid1 = "A-Za-z_'\u03b1-\u03ba\u03bc-\u03fb\u1f00-\u1ffe\u2070-\u2079"
+      name_mid2 = "\u207f-\u2089\u2090-\u209c\u2100-\u214f0-9."
+      name = %r([#{name_start}][#{name_mid1}#{name_mid2}]*)
+
+      state :root do
+        rule /\s+/m, Text::Whitespace
+        rule %r(/-), Comment::Multiline, :comment
+        rule /--.*$/, Comment::Single
```
Comment (pyrmont) [inline lib/rouge/lexers/lean.rb]: The `$` here is unnecessary, I think. `.*` will match to the end of the line.

File: lib/rouge/lexers/lean.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        rule /\"/, Str, :string
```
Comment (pyrmont) [inline lib/rouge/lexers/lean.rb]: Is there a reason to escape `"` here?

File: lib/rouge/lexers/lean.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        rule /\d+/, Num::Integer
+        rule /[~?][a-z][\w\']*:/, Name::Variable
```
Comment (pyrmont) [inline lib/rouge/lexers/lean.rb]: Is there a reason to escape `'` here? And it's correct that variables can have one or more `'` characters appearing at any point in the name? Including at the end?

File: lib/rouge/lexers/lean.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      state :comment do
+        rule %r([^/-]+), Comment::Multiline
+        rule(%r(/-)) { token Comment::Multiline; push }
+        rule %r(-/), Comment::Multiline, :pop!
+        rule /-/, Comment::Multiline
+      end
+
+      state :string do
+        rule /[^\\\"]+/, Str
+        rule /\\[n\"\\]/, Str::Escape
```
Comment (pyrmont) [inline lib/rouge/lexers/lean.rb]: So the only escaped characters are `\n`, `\"` and `\\`?

File: lib/rouge/lexers/lean.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        rule /\"/, Str, :pop!
```
Comment (pyrmont) [inline lib/rouge/lexers/lean.rb]: Again, don't think it's necessary to escape `"` here.

File: lib/rouge/lexers/lean.rb [hunk 7]
```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class Lean < RegexLexer
+      title "lean"
+      desc 'Lean (leanprover.github.io)'
+      tag 'lean'
+      mimetypes 'text/x-lean'
```
Comment (pyrmont) [inline lib/rouge/lexers/lean.rb]: There is no agreed upon file extension for Lean files?

**Comment by ice1000** [comment]
Is this ever gonna get merged?

**Comment by pyrmont** [comment]
@ice1000 There are a number of outstanding questions awaiting a response. I'm waiting on answers to those.

**Comment by hrldcpr** [comment]
#1798 is a more recent attempt at adding Lean support (which would be great btw!)

**Comment by pcapriotti** [comment]
This is based on an old version of lean. Closing in favour of #1798.

**Pull request closed without merging**